gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/vertx/web/handler/impl/internal/mfa/filter/StepUpAuthenticationFilter.java

```diff
@@ -38,22 +39,21 @@ public StepUpAuthenticationFilter(
 
     @Override
     public Boolean get() {
-        // If Adaptive MFA is active and KO (rule == false) we bypass this filter
-        // Because it could return true and skip MFA
-        if (!context.isMfaSkipped() && context.isAmfaActive() && !context.isAmfaRuleTrue()) {
-            return false;
-        }
         String mfaStepUpRule = context.getStepUpRule();
         if (context.isStepUpActive()) {
             context.setStepUpRuleTrue(isStepUpAuthentication(mfaStepUpRule));
         }
-        return !context.isAmfaActive() &&
-                context.isStepUpActive() &&
-                !context.isStepUpRuleTrue() &&
-                (context.isUserStronglyAuth() || context.isMfaSkipped());
+
+        // if strongly auth and StepUp rule is false (SAFE),
+        // then we can bypass MFA Challenge
+        if (context.isUserStronglyAuth()) {
+            return context.isStepUpActive() && !context.isStepUpRuleTrue();
```

Comment:
```suggestion
            return Boolean.FALSE.equals(context.isStepUpActive()) || Boolean.FALSE.equals(context.isStepUpRuleTrue());
```

Reply:
why do you propose this syntax when the both method are returning primitive value?

Reply:
it can be
context.isStepUpActive() == false || context.isStepUpRuleTrue() == false

if we have stepUp disabled we always get UNSAFE. We have to get Safe instead.

Reply:
ok right. 

However, context.isUserStronglyAuth() should be checked after the step up.

Reply:
(context.isUserStronglyAuth() || context.isMfaSkipped())
&& context.isStepUpActive() 
&& !context.isStepUpRuleTrue()